### PR TITLE
Improve output for 'chalice deploy'

### PR DESCRIPTION
**Pull request by jamesls**
This adds more detail to the output of `chalice deploy`, including:

* Add resource names where possible (lambda function names, rest api IDs, etc)
* Print when we're deleting unreferenced lambda functions

I've also introduced a module level logger that includes debug messages for the parts of the deployer.  This is more intended for us to help troubleshoot any deployment issues.

I've also restructured the internals to remove bare print calls.  A `UI` object is now passed through the deployer objects.  This makes testing easier as well as makes it easier to add more UI capabilities in the future.  Right now it just has writing to stdout/stderr, and prompting.  I've also moved the `NoPrompt` interface into the `UI` interface.

### CODE REVIEW

**Comment by jamesls** [comment]
If I'm understanding codecov correctly, it's complaining that I changed some of the print statements that were previously untested to `self._ui.write`?

File: chalice/deploy/packager.py [hunk 1]
```diff
@@ -87,7 +84,8 @@ def create_deployment_package(self, project_dir, python_version,
             except MissingDependencyError as e:
                 missing_packages = '\n'.join([p.identifier for p
                                               in e.missing])
-                print(MISSING_DEPENDENCIES_TEMPLATE % missing_packages)
+                self._ui.write(
+                    MISSING_DEPENDENCIES_TEMPLATE % missing_packages)
```
Comment (stealthycoin) [inline chalice/deploy/packager.py]: Need a `\n` here?

**Comment by stealthycoin** [comment]
Yes thats what it looks like its complaining about to me.

File: chalice/deploy/packager.py (hunk 1 above)
Comment (jamesls) [inline chalice/deploy/packager.py]: I cheated and updated the `MISSING_DEPENDENCIES_TEMPLATE` with a newline.

**Comment by jamesls** [comment]
In which case I'm 0 on adding tests for this, given the upcoming refactor to the deployer...

**Comment by stealthycoin** [comment]
Yea I think its fine to ignore for now and fix when you do something more substantial than change `print` to `ui.write`

File: chalice/deploy/packager.py (hunk 1 above)
Comment (stealthycoin) [inline chalice/deploy/packager.py]: Ah I missed that change.

**Pull request merged**